### PR TITLE
New keyboard cgame ABI

### PR DESCRIPTION
**Pull request by slipher**
This updates the cgame ABI to use the new keyboard key representations.

See https://github.com/Unvanquished/Unvanquished/pull/1065 for the cgame side.

### CODE REVIEW

File: src/engine/client/cl_cgame.cpp [hunk 1]
```diff
@@ -1474,74 +1447,65 @@ void CGameVM::QVMSyscall(int index, Util::Reader& reader, IPC::Channel& channel)
...
+					result = Keyboard::ScancodeToAscii(scancode);
+				}
 			});
 			break;
 
 		case CG_KEY_SETBINDING:
-			IPC::HandleMsg<Key::SetBindingMsg>(channel, std::move(reader), [this] (int keyNum, int team, std::string cmd) {
-				Keyboard::SetBinding(Keyboard::Key::FromLegacyInt(keyNum), team, std::move(cmd));
+			IPC::HandleMsg<Keyboard::SetBindingMsg>(channel, std::move(reader), [this] (Keyboard::Key key, int team, std::string cmd) {
+				Keyboard::SetBinding(key, team, std::move(cmd));
```
Comment (Kangz) [inline src/engine/client/cl_cgame.cpp]: Is any validation necessary for key?

File: src/engine/client/cl_cgame.cpp [hunk 2]
```diff
@@ -1474,74 +1447,65 @@ void CGameVM::QVMSyscall(int index, Util::Reader& reader, IPC::Channel& channel)
...
 				Key_ClearStates();
 			});
 			break;
 
 		case CG_KEY_KEYSDOWN:
-			IPC::HandleMsg<Key::KeysDownMsg>(channel, std::move(reader), [this] (std::vector<int> keys, std::vector<int>& list) {
+			IPC::HandleMsg<Keyboard::KeysDownMsg>(channel, std::move(reader), [this] (std::vector<Keyboard::Key> keys, std::vector<bool>& list) {
 				list.reserve(keys.size());
-				for (unsigned i = 0; i < keys.size(); ++i)
+				for (Keyboard::Key key : keys)
```
Comment (Kangz) [inline src/engine/client/cl_cgame.cpp]: ditto

File: src/common/Serialize.h [hunk 3]
```diff
@@ -31,7 +31,18 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #ifndef COMMON_SERIALIZATION_H_
 #define COMMON_SERIALIZATION_H_
 
+#include <limits>
```
Comment (Kangz) [inline src/common/Serialize.h]: These should already be included in common/Common.h no?

File: src/engine/client/key_binding.cpp [hunk 4]
```diff
@@ -288,12 +288,15 @@ Util::optional<std::string> GetBinding(Key key, BindTeam team, bool useDefault)
...
+std::vector<Key> GetKeysBoundTo(int team, Str::StringRef command)
 {
+	if (!IsValidTeamNumber(team)) {
+		return {};
+	}
 	std::vector<Key> result;
 	for (const auto& key : keys) {
-		for (BindTeam team : {GetTeam(), BIND_TEAM_DEFAULT}) {
-			const auto& binding = key.second.binding[team];
+		for (BindTeam t: {Util::enum_cast<BindTeam>(team), BIND_TEAM_DEFAULT}) {
```
Comment (Kangz) [inline src/engine/client/key_binding.cpp]: Needs validation?

File: src/engine/client/cl_cgame.cpp (hunk 1 above)
Comment (slipher) [reply src/engine/client/cl_cgame.cpp]: Good idea, I added some validation here as well as in the deserialization.

File: src/engine/client/cl_cgame.cpp (hunk 2 above)
Comment (slipher) [reply src/engine/client/cl_cgame.cpp]: `IsValid` is called in `IsDown` (though maybe `IsNonnull` or something like that would be a more accurate name).

File: src/common/Serialize.h (hunk 3 above)
Comment (slipher) [reply src/common/Serialize.h]: Well, we don't seem to use the precompiled header in cgame or sgame, so I can't assume everything is included.

File: src/engine/client/key_binding.cpp (hunk 4 above)
Comment (slipher) [reply src/engine/client/key_binding.cpp]: Of what?

**Comment by slipher** [comment]
@Kangz does it look OK now?

**Pull request merged**